keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,527 @@
+
+Namespaced objects will be represented by joining their name and namespace with `/` character. Pattern: `<namespace>/<name>`, example for pod: `kube-system/kubedns`
+
+Non-namespaced objects will be represented by their name. Pattern: `<name>`, example for node: `cluster1-vm-72x33b8p-34jz`
+
+Example:
+```go
+pod := corev1.Pod{Name: "kubedns", Namespace: "kube-system", ...}
+klog.InfoS("Pod status updated", "pod", pod, "status", "ready")
+```
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+```
+Selecting key under which put objects will be left to caller as defining log schema will need more discussion and is outside of scope of this KEP.
+
+### Selecting most important logs
+
+As migration to new message structure will be done manually we will be focusing on logs that have the greatest impact for log querying and processing. We will be focusing on log messages Proposed plan of measurements:
+
+* Using log data from single run of kubernetes [gce-master-scale-performance](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-master-scale-performance) tests
+* Based on logs generated on master node
+* Aggregate from components:
+  * kube-controller-manager
+  * kube-scheduler
+  * kube-apiserver
+  * kubelet
+* Only for log lines generated by klog (e.g. prefix `I0129 03:26:19.857798       1 garbagecollector.go:517]`)
+* Identified by code line that generated log (e.g. `garbagecollector.go:517`)
+
+We assume that exact details of taking measurements can be improved, but overall methodology is solid. With those criteria we concluded that covering 99.9% of logs on master node are generated by 22 log messages. Exact list of those logs is provided in the detail design section.
+
+#### JSON output format
+
+Introduction of new methods to klog library will make identification of different components of the log message much easier.
+With klog v2 we can take further advantage of this fact and add an option to produce structured logs in JSON format.
+
+Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Proposed default fields:
+* `ts` - timestamp as Unix time (required, `float64`)
+* `v` - verbosity (required, `int`, default 4)
+* `err` - error string (optional, `string`)
+* `msg` - message (required, `string`)
+
+Default Serialization:
+* primitive types (`int`, `float`) - Compatible with [golang json package](https://golang.org/pkg/encoding/json/#Marshal)
+* namespaced Kubernetes objects (`Pod`, `Deployment`) - JSON object consisting of `name` and `namespace`. Example `{"name": "kubedns", "namespace": "kube-system"}`
```

Comment:
this isn't enough to be able to track back which object is failing.  UID and resourceversion are important when present.

Reply:
Currently logs don't use UID and resourceversion. Let's not try to add every possible metadata at once.
I would welcome adding UID, but for some it could be to big increase of log volume. 

Proposing a schema, adding optional fields (e.g. UID you proposed) and versioning should be covered on next step.

Reply:
Uid is useful kubelet side when debugging pods with same name and namespace that are short lived.